### PR TITLE
ref(invite-members): refactor invite members modal

### PR DESCRIPTION
**Pull request by ameliahsu**
refactor invite members modal to avoid prop drilling 👷🏼‍♀️

### CODE REVIEW

File: static/app/components/modals/inviteMembersModal/inviteMembersContext.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  setEmails: (emails: string[], index: number) => void;
+  setRole: (role: string, index: number) => void;
+  setTeams: (teams: string[], index: number) => void;
+  willInvite: boolean;
+  error?: string;
+};
+
+export const InviteMembersContext = createContext<InviteMembersContextProps | undefined>(
+  undefined
+);
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteMembersContext.tsx:26]: Rather than using `undefined` here, it's more common practice to provide a skeleton / loading state of the context - where things are either in "pending" or "off" states.
```suggestion
export const InviteMembersContext = createContext<InviteMembersContextProps>({
  complete: false,
  inviteStatus: {},
  invites: [],
  pendingInvites: [],
  reset: () => {},
  sendInvites: () => {},
  sendingInvites: false,
  setEmails: () => {},
  setRole: () => {},
  setTeams: () => {},
  willInvite: false,
});
```

This ends up helping with the type definitions as well - it helps point out if things are optional or not and what the initial states should look like.

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+export default function InviteMembersFooter({canSend}: Props) {
+  const {willInvite, invites, sendInvites, inviteStatus, sendingInvites, complete} =
+    useContext(InviteMembersContext) ?? {
+      willInvite: false,
+      invites: [],
+      sendInvites: () => {},
+      inviteStatus: {},
+      sendingInvites: false,
+      complete: false,
+    };
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteMembersFooter.tsx:23]: if you change the initialization you should be able to remove the `?? { ... }`

File: static/app/components/modals/inviteMembersModal/inviteRowControlNew.spec.tsx [hunk 3]
```diff
@@ -25,74 +25,108 @@ describe('InviteRowControlNew', function () {
...
+            role: role || '',
+          },
+        ],
+        reset: () => {},
+        sendInvites: () => {},
+        sendingInvites: false,
+        setEmails: mockOnChangeEmails,
+        setRole: mockOnChangeRole,
+        setTeams: mockOnChangeTeams,
+        willInvite: true,
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteRowControlNew.spec.tsx:47]: can simplify these tests to only what you want to test and allow the context defaults to set everything else.

File: static/app/components/modals/inviteMembersModal/inviteRowControlNew.tsx [hunk 4]
```diff
@@ -43,19 +35,28 @@ function mapToOptions(values: string[]): SelectOption[] {
...
+  const {inviteStatus, pendingInvites, setEmails, setRole, setTeams, reset} = useContext(
+    InviteMembersContext
+  ) ?? {
+    inviteStatus: {},
+    pendingInvites: [],
+    setEmails: () => {},
+    setRole: () => {},
+    setTeams: () => {},
+    reset: () => {},
+  };
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteRowControlNew.tsx:48]: 👋

**Review by saponifi3d — COMMENTED** [review]
just comments on the context initialization (and all the 🔴 we can gain) i'll take another pass in the am :)

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        complete: true,
+        inviteStatus: {
+          'moo-deng': {sent: false, error: 'Error'},
+          'moo-waan': {sent: false, error: 'Error'},
+        },
+        willInvite: true,
+      })
+    );
+    expect(screen.getByText(/sent , 2 failed to send\./i)).toBeInTheDocument();
+    expect(screen.getByText(/0 invites/i)).toBeInTheDocument();
```
Comment (ameliahsu) [inline static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx:89]: a little janky but had to test it this way because the sent invite count is in a separate `<strong>` component 🫠

File: static/app/components/modals/inviteMembersModal/index.tsx [hunk 6]
```diff
@@ -65,6 +71,22 @@ function InviteMembersModal({
...
+
+  const inviteMessage = willInvite ? (
+    <Subtext>{t('Invite unlimited new members to join your organization.')}</Subtext>
+  ) : (
+    <Alert type="warning" showIcon>
+      {t(
+        'You can’t invite users directly, but we’ll forward your request to an org owner or manager for approval.'
+      )}
+    </Alert>
+  );
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/index.tsx:88]: nit these could be separate components with props. the advantage of having this defined as a separate component is how JS will run this code.

in practice what will happen is that every time there's a state change or button click, we'll redefine this component. if we pull it out as a separate component - we will only need to recompute these if the property changes.

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx (hunk 5 above)
Comment (saponifi3d) [reply static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx:89]: hm - may just want to target that tag and just check the number in that case? i tend to not like targeting tests with `text` as that's not as hardcoded as you'd always think (like if we wanted to run tests in multiple languages to ensure everything is translated, these would fail). i normally add a `data-` attribute to be really specific with my targeting, if i can't use the other selectors to find something.

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      })
+    );
+    expect(screen.getByText(/sent , 1 failed to send\./i)).toBeInTheDocument();
+    expect(screen.getByText(/1 invite/i)).toBeInTheDocument();
+  });
+
+  it('displays correct status message for multiple sent and failed invites', function () {
+    render(
+      getComponent({
+        ...defaultInviteProps,
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx:111]: i feel like this read really nicely for the tests 🎉 - it's really clear the settings going into the component and the expected results. 🎉

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.tsx [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+export default function InviteMembersFooter({canSend}: Props) {
+  const {complete, inviteStatus, invites, sendInvites, sendingInvites, willInvite} =
+    useContext(InviteMembersContext);
+  const isValidInvites = invites.length > 0;
+
+  return (
+    <FooterContent>
+      <div>
+        {/* TODO(mia): remove these props and use InviteMemberContext once old modal is removed */}
```
Comment (saponifi3d) [inline static/app/components/modals/inviteMembersModal/inviteMembersFooter.tsx:22]: if you wanted.. you could move the context to wrap both components - then refactor these as well. (if you wanted to do that, it'd probably be better as a separate PR tho)

**Review by saponifi3d — APPROVED** [review]
🎉 this refactor looks awesome. thanks for diving into everything!

File: static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx (hunk 5 above)
Comment (ameliahsu) [reply static/app/components/modals/inviteMembersModal/inviteMembersFooter.spec.tsx]: update: fixed this for successfully sent invites, but still using the text "failed" to look for the failed invite count. i plan on changing the wording of the status message in my next PR, so this will get fixed 👍
